python/ray/data/dataset_pipeline.py

```diff
@@ -1150,12 +1150,36 @@ def iter_datasets(self) -> Iterator[Dataset[T]]:
         if self._executed[0]:
             raise RuntimeError("Pipeline cannot be read multiple times.")
         self._executed[0] = True
-        if self._first_dataset is None:
-            self._peek()
-        iter = itertools.chain([self._first_dataset], self._dataset_iter)
-        self._first_dataset = None
-        self._dataset_iter = None
-        return iter
+
+        # If the first dataset has already been executed (via a peek operation), then
+        # we don't re-execute the first dataset when iterating through the pipeline.
+        # We re-use the saved _first_dataset and _remaining_dataset_iter
+        if self._first_dataset is not None:
+
+            class _IterableWrapper(Iterable):
```

Comment:
Is this wrapping needed since iterator itself has __iter__ to return itself?

Reply:
+1 to this question @amogkam, I don't think that we need this wrapper, do we?